apps/opik-backend/src/main/java/com/comet/opik/domain/llmproviders/OpenAi.java

```diff
@@ -0,0 +1,108 @@
+import jakarta.inject.Inject;
+import jakarta.ws.rs.ClientErrorException;
+import jakarta.ws.rs.InternalServerErrorException;
+import jakarta.ws.rs.ServerErrorException;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+
+@Slf4j
+public class OpenAi implements LlmProviderService {
+    private static final String UNEXPECTED_ERROR_CALLING_LLM_PROVIDER = "Unexpected error calling LLM provider";
+
+    private final LlmProviderClientConfig llmProviderClientConfig;
+    private final OpenAiClient openAiClient;
+
+    @Inject
+    public OpenAi(LlmProviderClientConfig llmProviderClientConfig, String apiKey) {
+        this.llmProviderClientConfig = llmProviderClientConfig;
+        this.openAiClient = newOpenAiClient(apiKey);
+    }
+
+    @Override
+    public ChatCompletionResponse generate(@NonNull ChatCompletionRequest request, @NonNull String workspaceId) {
+        log.info("Creating chat completions, workspaceId '{}', model '{}'", workspaceId, request.model());
+        ChatCompletionResponse chatCompletionResponse;
+        try {
+            chatCompletionResponse = openAiClient.chatCompletion(request).execute();
+        } catch (RuntimeException runtimeException) {
+            log.error(UNEXPECTED_ERROR_CALLING_LLM_PROVIDER, runtimeException);
+            if (runtimeException.getCause() instanceof OpenAiHttpException openAiHttpException) {
+                if (openAiHttpException.code() >= 400 && openAiHttpException.code() <= 499) {
+                    throw new ClientErrorException(openAiHttpException.getMessage(), openAiHttpException.code());
+                }
+                throw new ServerErrorException(openAiHttpException.getMessage(), openAiHttpException.code());
+            }
+            throw new InternalServerErrorException(UNEXPECTED_ERROR_CALLING_LLM_PROVIDER);
+        }
+        log.info("Created chat completions, workspaceId '{}', model '{}'", workspaceId, request.model());
+        return chatCompletionResponse;
+    }
+
+    @Override
+    public void generateStream(
+            @NonNull ChatCompletionRequest request,
+            @NonNull String workspaceId,
+            @NonNull Consumer<ChatCompletionResponse> handleMessage,
+            @NonNull Runnable handleClose,
+            @NonNull Consumer<Throwable> handleError) {
+        log.info("Creating and streaming chat completions, workspaceId '{}', model '{}'", workspaceId, request.model());
+        openAiClient.chatCompletion(request)
+                .onPartialResponse(handleMessage)
+                .onComplete(handleClose)
+                .onError(handleError)
+                .execute();
+        log.info("Created and streaming chat completions, workspaceId '{}', model '{}'", workspaceId, request.model());
+    }
+
+    @Override
+    public ErrorMessage mapError(Throwable throwable) {
```

Comment:
this can probably be generalized

Reply:
Please look at the implementation. It uses openai specific code.